# Cancer Genomics Course

---
**University of Sydney - 25th to 27th July 2017**

## Instructors

* Katherine Champ (KC)- Bioplatforms Australia, Sydney
* Ann-Marie Patch (AMP) - QIMR Berghofer, Brisbane
* Gayle Philip (GP) - Melbourne Bioinformatics, Melbourne
* Erdahl Teber (ET) - CMRI, Sydney
* Sonika Tyagi (ST) - Monash University, Melbourne


## Timetable
### Day 1
**25th July** - [Computer Lab 1.4, Charles Perkins Centre, University of Sydney, NSW](https://goo.gl/maps/feoXLjEhke52)

| **Time** | **Topic** | **Links** | **Instructor** |
| -------- | --------- | --------- | ----------- |
|9:00|Welcome||KC|
|9:15|Introduction to cancer genomics and NGS techniques – focus on DNA||AMP|
|10:15|**Break**|
|10:30|Experimental design (interactive/ice breaker/group activity) and caveats; considerations on processing capacities; 10 ways to ruin your experiment||AMP|
|11:30|Command line intro (Unix, R) – L(15') / P(30')|[[Practical]](https://bpa-csiro-workshops.github.io/cancer-manuals/modules/cancer-module-cli/commandline/)|GP|
|12:30|**Lunch**|
|13:30|Raw data - FASTQ format and QC||ST|
|14:00|Alignment (L)||ST|
|14:30|Practical: Manipulation of BAM files and QC|[[QC Practical]](https://bpa-csiro-workshops.github.io/cancer-manuals/modules/cancer-module-qc/ngs-qc/)[[Alignment Practical]](https://bpa-csiro-workshops.github.io/cancer-manuals/modules/cancer-module-alignment/alignment/)|ST/GP|
|15:00|**Coffee break**|
|15:15|Practical: Manipulation of BAM files and QC (cont.)||GP/ST|
|17:00|Q&A||All|

### Day 2
**26th July** - [Computer Lab 1.4, Charles Perkins Centre, University of Sydney, NSW](https://goo.gl/maps/feoXLjEhke52)

| **Time** | **Topic** | **Links** | **Instructor** |
| -------- | --------- | --------- | ----------- |
|9:00|SNV detection (review, germline vs somatic, tools, pitfalls, data visualization) – L/P. Indels – cover in SNV lecture + bonus exercises (P), specific challenges of indels analysis, tools||ST|
|9:45|Practical: SNV detection|[[Practical]](https://bpa-csiro-workshops.github.io/cancer-manuals/modules/cancer-module-snv/snv/)|ST|
|10:30|**Coffee break**|
|10:45|Variants annotation and filtration (L) – tools landscape||ST|
|11:00|Practical: Variants visualization (IGV), annotation and filtration|[[Practical]](https://bpa-csiro-workshops.github.io/cancer-manuals/modules/cancer-module-snv/snv/)|ST|
|12:30|**Lunch + coffee**|
|13:30|CNV analysis using NGS data (L)||AMP|
|14:15|Practical: CNV analysis – deletion/amplification, calling CNVs, visualization, interpretation|[[Practical]](https://bpa-csiro-workshops.github.io/cancer-manuals/modules/cancer-module-cnv/cnv-tut/)|GP/AMP|
|15:00|**Break**|
|15:15|Practical: CNV analysis – deletion/amplification, calling CNVs, visualization, interpretation (cont.)||GP/AMP|
|17:00|Q&A||All|

### Day 3
**27th July** - [Computer Lab 1.4, Charles Perkins Centre, University of Sydney, NSW](https://goo.gl/maps/feoXLjEhke52)

| **Time** | **Topic** | **Links** | **Instructor** |
| -------- | --------- | --------- | ----------- |
|9:00|SV analysis – breakpoints/fusion, tools||AMP|
|9:45|Practical: SV analysis – breakpoints/fusion|[[Practical]](https://bpa-csiro-workshops.github.io/cancer-manuals/modules/cancer-module-sv/sv_tut/)|ET|
|10:30|**Coffee break**|
|10:45|Practical: SV analysis – breakpoints/fusion (cont.)||ET|
|12:30|**Lunch**|
|13:30|Downstream analysis and interpretation – Exploration of resources that can be used for this. E.g. databases (COSMIC, TCGA, etc.), integration with clinical information|[[Practical 1]](https://bpa-csiro-workshops.github.io/cancer-manuals/modules/cancer-module-somatic/01_signatures/) [[Practical 2]](https://bpa-csiro-workshops.github.io/cancer-manuals/modules/cancer-module-somatic/02_intogen/)|AMP|
|14:15|Practical: Downstream analysis and interpretation||AMP|
|15:00|**Coffee break**|
|15:15|Practical: Downstream analysis and interpretation(cont.)||AMP|
|16:00|How does it all link together? Integration of different data types||Mark Cowley|
|16:45|Q&A, wrap up (how to access course’s VM) & survey||All|
